Feature: Setup dotenv

antonykamp opened the pull request:
Fixes #67 

Install dev dependency `dotenv` to load env variables from files.
Add constants file to test and src folders. These contain environment variables in the future.
Add .env.secret.sample file to define needed env variables.

You can find the edited wiki entry here:
https://github.com/BP2022-AP1/bp2022-ap1/wiki#environment-variables

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [ ] ~Breaking changes anounced~
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [x] Tests and linter have passed AFTER local merge
- [x] Another dev reviewed and approved

antonykamp committed:
dep: install python-dotenv

Install dev dependency dotenv to load env variables from files.

antonykamp committed:
feat: setup dotenv

Add constants file to test and src folders. These contain environment variables in the future.
Add .env.secret.sample file to define needed env variables.

antonykamp committed:
fix: lint

antonykamp committed:
fix: import order

antonykamp committed:
fix: import order

antonykamp committed:
doc: add information about usage of environment variables to README

antonykamp committed:
doc: replace explanation with link to wiki

SaturnHafen commented on src/constants.py:
```diff
@@ -0,0 +1,9 @@
+
+from dotenv import dotenv_values
+
+constants = {
+    **dotenv_values(".env.shared"),  # load shared development variables
+    **dotenv_values(".env.secret"),  # load sensitive variables
+    **os.environ,  # override loaded values with environment variables
+}
```
Should we add a test to check if an environment variable appears in the constants dict?

antonykamp replied:
In my opinion, tests should test the functionality our program, not the development environment. What do you think?

SaturnHafen replied:
Sounds reasonable. An other issue that might be worth testing for: The preference of loaded env-vars (`os.environ` is more important than `env.secret` is more important than `env.shared`; in case someone restructures this code later).

antonykamp replied:
That sounds reasonable 👍 I'll add it right away

SaturnHafen reviewed (approved):
We should probably add an example to the wiki regarding the use of the `constants.py` file. Additionally, I fixed some spelling mistakes in the wiki.

antonykamp commented:
Thanks for fixing some spelling in the wiki 🙌 I can add an example to the wiki, that's a great idea!

SaturnHafen commented on pyproject.toml:
```diff
@@ -22,6 +22,7 @@ pytest = "^7.2.2"
 black = "^23.1.0"
 isort = "^5.12.0"
 pylint = "^2.17.0"
+python-dotenv = "^1.0.0"
```
You declared dotenv as dev-dependency, shouldn't it be a normal dependency instead?

antonykamp replied:
It's a dev dependency, because we will use it locally only. It doesn't brings functionality to the user.

SaturnHafen replied:
Fine, but then this import will break in deployment, or am I mistaken?
<https://github.com/BP2022-AP1/bp2022-ap1/pull/70/files#diff-42d0f00c4780305311eae71fecca993a0523587cb4156e039cabd0d1ad8dbbb8R3>

antonykamp replied:
I'm not a poetry-export, but in node dev-dependencies will be installed in production too. That's why I thought, that`poetry install` installs every dependency registered in the `pyproject.toml`. The groups are just a help for the developer.

antonykamp committed:
feat: add type declaration

Add type declaration to constants dictionary in constants.py files.

antonykamp committed:
feat: add tests

Add tests to verify the loading order of the environment files for the development and test environment.

antonykamp committed:
fix: lint

Disable pylint redefined-outer-name and unused-argument for some lines

antonykamp committed:
fix: lint

Disable import-outside-toplevel for some lines

antonykamp committed:
fix: lint

antonykamp committed:
Merge branch 'dev' into 67-setup-dotenv
